piptools/scripts/compile.py

```diff
@@ -545,27 +630,41 @@ def cli(
             constraints.extend(reqs)
         elif is_setup_file:
             setup_file_found = True
+            src_dir = os.path.dirname(os.path.abspath(src_file))
             try:
                 metadata = project_wheel_metadata(
-                    os.path.dirname(os.path.abspath(src_file)),
+                    src_dir,
                     isolated=build_isolation,
                 )
             except BuildBackendException as e:
                 log.error(str(e))
                 log.error(f"Failed to parse {os.path.abspath(src_file)}")
                 sys.exit(2)
 
-            constraints.extend(
-                parse_requirements_from_wheel_metadata(
-                    metadata=metadata, src_file=src_file
+            if not build_deps_only:
+                constraints.extend(
+                    parse_requirements_from_wheel_metadata(
+                        metadata=metadata, src_file=src_file
+                    )
+                )
+                if all_extras:
+                    if extras:
+                        msg = "--extra has no effect when used with --all-extras"
+                        raise click.BadParameter(msg)
+                    extras = tuple(metadata.get_all("Provides-Extra"))
+            if build_deps_for_distributions:
+                package_name = metadata.get_all("Name")[0]
+                constraints.extend(
+                    _build_requirements(
+                        # Build requirements will only be present if a pyproject.toml file exists,
+                        # but if there is also a setup.py file then only that will be explicitly
+                        # processed due to the order of `DEFAULT_REQUIREMENTS_FILES`.
+                        src_dir,
+                        "pyproject.toml",
+                        build_deps_for_distributions,
+                        package_name,
+                    )
```

Comment:
On line 635, we call `project_wheel_metadata()`, which installs `build_system_requires` in an isolated environment. Consequently, `_build_requirements()` installs `build_system_requires` for a second time in a different isolated environment, leading to a performance impact. Could we fix the code to build wheel metadata and build requirements only once?

Reply:
Addressed in https://github.com/apljungquist/pip-tools/pull/1

Reply:
I'll mark this as resolved since the PR has been merged.